src/storage/engine/mod.rs

```diff
@@ -328,13 +329,17 @@ impl<I: Iterator> Cursor<I> {
             scan_mode: mode,
             min_key: None,
             max_key: None,
+            valid: true,
         }
     }
 
     pub fn seek(&mut self, key: &Key, statistics: &mut CFStatistics) -> Result<bool> {
+        self.valid = true;
+
         assert_ne!(self.scan_mode, ScanMode::Backward);
         if self.max_key.as_ref().map_or(false, |k| k <= key.encoded()) {
             self.iter.validate_key(key)?;
```

Comment:
seem must set valid before this?

Reply:
The cursor may be in a `self.valid == false` state, for example, we have data range [5, 10] and there are following operations:

1. seek 15 (upper bound = 15, self.valid == true, iter.valid == false)
2. seek 16 (upper bound == 15, self.valid == false, iter.valid == false)
3. seek 5 (self.valid == true, iter.valid == true)

after operation 3, self.valid should be changed.

Reply:
The logic of this PR is very tricky. Maybe TLA+ can help us prove that every corner case is covered?

Reply:
confusing......